### PR TITLE
SONARJAVA-2686 Rule S4508 Object deserialization from untrusted source

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S4508_java.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<p>Object deserialization from an untrusted source can lead to unexpected code execution. ObjectInputStream doesn't provide a way to apply rules on
+its InputStream argument. Knowing that all serializable classes in the scope of the classloader will be deserialized, there is a possibility to
+execute malicious code during the deserialization phase even if, at the end, a ClassCastException will be raised.</p>
+<p>This rule raises an issue when one of the methods "readArray", "readObject", "readUnshared" is called on a "java.io.ObjectInputStream" or any
```
Comment (andrei-epure-sonarsource) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S4508_java.html]: please update rule to not contain "readArray"

File: java-checks/src/main/java/org/sonar/java/checks/ObjectDeserializationCheck.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+@Rule(key = "S4508")
+public class ObjectDeserializationCheck extends AbstractMethodDetection {
+
+  private static final String OBJECT_INPUT_STREAM = "java.io.ObjectInputStream";
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(
+      MethodMatcher.create().typeDefinition(OBJECT_INPUT_STREAM).name("readObject").withoutParameter(),
```
Comment (andrei-epure-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/ObjectDeserializationCheck.java]: does it make sense to check for subclasses of `ObjectInputStream`, and also look for `readObjectOverride​`?

File: java-checks/src/test/files/checks/ObjectDeserialization.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package test;
+
+import java.io.ObjectInputStream;
+
+public class ObjectDeserialization {
+
+  public void foo() {
+    InputStream untrusted = new WhatEverInputStream();
+    ObjectInputStream ois = new ObjectInputStream(untrusted);
+    Foo deserialized = (Foo) ois.readObject();  // Noncompliant {{Verify there is no risk to deserialize from this source}}
```
Comment (andrei-epure-sonarsource) [inline java-checks/src/test/files/checks/ObjectDeserialization.java]: nitpick: specify the highlighting coordinates for the first message

File: java-checks/src/main/java/org/sonar/java/checks/ObjectDeserializationCheck.java (hunk 2 above)
Comment (saberduck) [reply java-checks/src/main/java/org/sonar/java/checks/ObjectDeserializationCheck.java]: Subclasses will be handled by MethodMatcher automatically. I don't see why we should check for `readObjectOverride`, can you provide some reasoning?

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S4508_java.html (hunk 1 above)
Comment (saberduck) [reply java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S4508_java.html]: Done. Thanks!

File: java-checks/src/main/java/org/sonar/java/checks/ObjectDeserializationCheck.java (hunk 2 above)
Comment (saberduck) [reply java-checks/src/main/java/org/sonar/java/checks/ObjectDeserializationCheck.java]: I added test case with subclass

**Pull request merged**